### ISSUE TITLE
Reorganize annotations

### ISSUE THREAD

**Issue by Pazus** [utPLSQL/utPLSQL#133]
Lets reogranize all our annotations to match common approaches implemented in jUnit 5 and RSpec. This will close the framework to other testing fremawork out there and make learning easier for other programmers.

Here are guiding documents we should keep in minds:
NUnit 3: https://github.com/nunit/docs/wiki/Attributes
JUnit 5: http://junit.org/junit5/docs/current/user-guide/#writing-tests-annotations
RSpec : (@jgebal  please provide correct link)

| Annotation |Level| Describtion |
| --- | --- | --- |
| `%suite` | Package | Marks package to be a suite of tests This way all testing packages might be found in a schema. |
| `%suitepath(<path>)` | Package | Similar to java package. The annotation allows logical grouping of suites into hierarcies. |
| `%displayname(<description>)` | Package/procedure | Human-familiar describtion of the suite/test. `%displayname(Name of the suite/test)` |
| `%test` | Procedure | Denotes that a method is a test method. |
| `%beforeall` | Procedure | Denotes that the annotated procedure should be executed before all `%test` methods in the current suite. |
| `%afterall` | Procedure | Denotes that the annotated procedure should be executed after all `%test` methods in the current suite. |
| `%beforeeach` | Procedure | Denotes that the annotated procedure should be executed before each `%test` method in the current suite. |
| `%aftereach` | Procedure | Denotes that the annotated procedure should be executed after each `%test` method in the current suite. |
| `%beforetest(<procedure_name>)` | Procedure | Denotes that mentioned procedure should be executed before the annotated `%test` procedure. |
| `%aftertest(<procedure_name>)` | Procedure | Denotes that mentioned procedure should be executed after the annotated `%test` procedure. |
| `%rollback(<type>)` | Package/procedure | Configure transaction control behaviour (type). Supported values: `auto`(default) - rollback to savepoint (before the test/suite setup) is issued after each test/suite teardown; `manual` - rollback is never issued automatically. Property can be overridden for child element (test in suite) |
| `%disable` | Package/procedure | Used to disable a suite or a test |
| `%tag(<tag>)` | Package/procedure | Used to declare _tags_ for filtering tests, either at the suite or test level. |
| `%nested` | Package | Denotes the beginning of the nested suite section. All `%test` procedures, declared before the `%endnested` annotation are treated as procedures of the nested suite with their own `%beforeall`, `%afterall`, `%beforeeach`, `%aftereach` methods also declared in the block. Only one level of nesting is allowed. |
| `%endnested` | Package | Denotes the end of the nested suite section. |
| `%raise` (rejected) | Procedure | Denode that exception is expected as the result of the execution. The expected exception number is provider as `%raise(-20001)` |

Thought about the `%order` annotation, but I think that should be a runtime parameter.

**Comment by jgebal** [comment]
I really like the new concept.
Nunit seems much more complicated than Junit. Simple design is something to follow.
One thing in question is the annotation `%raise`.
I'm not sure how this would work.
Can you go ahead with changes excluding the `%raise`?
I'd like us to discuss the possible alternatives and options.

**Comment by Pazus** [comment]
Yes. It's optional as it is not yet implemented and currently not even in the scope. I wanted to share a variant how it might look like.

Though I think this kind of annotation should arise later. The similar thing exists in jUnit 4 and was replaced in jUnit 5 with lambdas. As we don't have one the old jUnit concept might be appropriate.

**Comment by Pazus** [comment]
And yes. NUnit on the first glance looks over complicated compared to jUnit. Even its 3rd version.

**Comment by rlove** [comment]
The proposal It seem very clean and easy to use.   

When using: `%beforeall`, `%afterall`, `%beforeeach`,  or `%aftereach`

What determines the order of execution multiple procedures are annotated with the same annotation value, or will it fail?    I am not sure we need to support multiples but I am just trying clarify ambiguities that may exist. 

I that same context for documentation purposes how many times can each annotation be used.
Then for testing purpose we should see how we behave when we exceed these limits and make sure it's easy to figure out what went wrong.

**Comment by jgebal** [comment]
Really good points @rlove
I would say that There Can be only one beforeAll, beforeEach and afterAll, afterEach within single context (suite or nested block).
BeforeAll is always executed first within a context,  then before each test t gr8 e beforeEach is executed, then the test, then afterEach and one all tests in context are executed then the afterAll is executed for context.

Oh. Now I realise it's much better to use `%context` and `%endContext` to do nesting within package. 
Context could have name using the`%displayname` annotation.
And I would really like to allow unlimited nesting of contexts.
@pazus I no-one I'm asking alot, but it's Christmas time and I was a really good boy the whole year ;)

**Comment by jgebal** [comment]
Oh, and two more things.
We no longer use `%suitepath` so we go back to stating that one package is one suite and that's it.
I like this idea as binding packages together brings a complexity and corner cases that are hard to handle (loops).

`%tag` should be accepted on `%test`, `%suite` and `%context`, as those are the elements that group expectations (assertions)

**Comment by rlove** [comment]
@jgebal How do we group test to be run together then?     We have version 2 suites that do not share a setup/teardown but are run as a group.

I can see the desire to not have a shared setup/teardown for the suite.   But to me a suite is set of tests that may span more than one packages.

**Comment by Pazus** [comment]
Thank you all for the feedback!

Let me clarify it a bit.
Proposed annotations mostly are simple renames for current set of annotations, logic stays the same.
@jgebal is right about the order of `%beforeall`, `%beforeeach`, `%aftereach`, `%afterall` procedures. If an annotation block has multiple occrancies of the same annotation, the last is used. (This might be changed in the future for tags for example). And currently there is no warning when a procedure is marked with several such annotations. (in the implementation there is a hidden priority-based approach).

I don't like the `%context` idea. Why, as the package is marked as `%suite`, the subpart of it, which is actually also a suite with all the same behavior, should be named "context"? This breaks naming approach I think. Maybe lets name it `%nestedsuite`... or just stay with `%nested`.

@jgebal I don't understand why you say we no longer use `%suitepath`. We do use it. `%suitepath` is a great way to form hierarcies of suites. We need to handle such conflicts and that's it. (Even Oracle can do it with nocycle :) )

`%tag` is a "Package/procedure" scoped annotation so it is usable for both suites and tests (subsuite is still a suite).

@rlove on grouping tests. `%suitepath` is the way to group them.
Imagine I have packages `foo1` and `foo2`, each is a separate suite with tests (marked as `%suite`), each have its own `%beforeall` etc procedures and tests.
But also each of these packages is annotation with `%suitepath(foos)` which mean that the fully qualified path of the suites are:
foos/foo1
foos/foo2

foos - is *another suite*. There might be no such package so it is a logical suite. Or it can be declared as a package `foos` annotated with `%suite`. It even might have `%beforeall`/`%afterall` procedures which are be execute before we start suites `foo1`,`foo2`.

This way we can construct suite hierarcy. Even more using currently implemented runner I can runn a single suite/test at the any level of the hierarcy (for example `ut3:foos:foo2:test3`). The runner will skill all other nodes except setups/teardowns and the defined node of the hierarcy. So in this case the execution will be:
`foos` beforeall
`foo2` beforeall
`foo2` beforeeach
`foo2` test3
`foo2` aftereach
`foo2` afterall
`foos` afterall

This is the cross-package suite grouping in version3. And it IS alreay implemented.


Finaly lets list the changes, that are to be made:
1. remove `%testsetup`, `%testteardown`
2. renames: 
`%setup` - `%beforeeach`
`%teardown` - `%aftereach`
`%suitesetup` - `%beforeall`
`%suiteteardown` - `%afterall`
`%suitepackage` - `%suitepath`
`%ignore` - `%disabled`
3. `%suite`, `%test` annotations no longer accept describtion of the suite/test, `%displayname` is used instead
4. implement `%tag`, `%nested`, `%endnested` annotations (another PR)
5. `%raise` is currently out of scope

**Comment by rlove** [comment]
I believe the rename is very clear way to go.

But I am concerned about some longer usability issues.   I can see someone copying the package spec of another package as a starting point.   They may forget to change the suite name.   

What will happen if we have two packages with the same suite name both with before/after calls?   

I think we may need another issue to deal with warning users of these situations, where only the last package found will execute. 

I am also not ok with silently picking a method to use when there is an ambiguity in the definition.  i.e. multiple before/after calls.     So I am thinking we need an issue for this.

**Comment by Pazus** [comment]
@rlove: we don't have suitename. `Displayname` is an additional string description of a test/suite, but it is not used either for identification nor suitepath generation. Remember %

You've mentioned a nice example. Imagine I have package `foo` with suitepath `api`. Than I copy-paste the package to make an additional one. I have to rename the new package right? Otherwise I'll overwrite the old one. So I have `foo2` package with the same content, same `displayname`, same `suitepath`. But the full qualified names of this packages are `<user>:api.foo1`, `<user>:api.foo2`.
The same thing is used in java: in a class I specify `package`, and then import it using `import <package_path>.classname;`. Remember that also you can import a subtype, declared in a type, so <package_path> can be purely logical, but could contain real classes.

So the idea behind is very similar.

**Comment by jgebal** [comment]
Hmmm...
Would the `%suitepath` need to contain full hierarchy of ancestors or point only to the direct parent?
If it is to contain only the direct parent maybe we should call it: `%parentsuite`?
Pointing direct parent seems a better choice.
Having full path brings maintenance burden.

**Comment by Pazus** [comment]
The `%suitepath` contains (exactly as id does corrently) "path" to the parent node. It doesn't contain the name or anything of the current suite as current suite's name is the package name. It is currently so.
Thу concept is very similar to java "package" in class definition. That's why it was `%suitepackage`. There is a chance that the old name is better...

P.S. I feel a bit strange discussing how the current implementation works with people involved in development of that implementation :)

**Comment by jgebal** [comment]
@pazus I think the original concept of `%suitepath` should be changed, so in any package we only point to direct parent not the whole path all the way to the root.
If we don't change it then whenever you want to reorganize your packages (rename or remove a package) you need to review whole tree of packages and update the `%suitepath` everywhere.
When using `%parent` you only need to update packages that directly reference the renamed/removed package.
The down side `%parent` vs `%suitepath` is that `%parent` doesn't give you information about full hierarchy.
The good thing is that it is easier to maintain.

Please let me know what you think and do you think that parent would be a bete option then the path/suitepath annotation.

**Comment by Pazus** [comment]
I don't agree with you, @jgebal.

The package suitepath prevents you from cycles, I have realized that recently, parent approach makes cycles frequent.

With suitepath you see the whole hierarchy as you said and if you need to reorganize it you just you cross-files find-and-replace. Lots of programs enable that. 
Remember jetbrain IDEs also make it the main way of refactoring, so it's up to IDE not the code to help you maintain/manipulate it. The goal of a code is to make it as clear as possible.
This approach is well proven by java. 
Mirroring folder structure with packages in java is only a convenient option, but not necessary.

Btw in real world cases the structure of suitepaths is going to be flat, I suppose;)

**Comment by Pazus** [comment]
Bump!

**Comment by Pazus** [comment]
I've tried to implement tests without using `testsetup` and it appeared to be not convinient. that way I have to explicitly handle exceptions to perform teardown regardless of the test results. Test code become longer and more complex, results due to reraises are not that nice.

I think we should add `%beforetest`, `%aftertest` annotations back. So the cleanup after the test is back to the framework.

**Comment by jgebal** [comment]
@Pazus I guess having `%beforetest` and `%aftertest` as an optional annotation is a good idea.
Developers will have freedom of choice. If they want to embed setup/teardown inside the test, they do not use the annotations.
If they want to separate them from the test itself then they use it.
If I have `%beforeall` and `%beforeeach` and `%beforetest` defined I would expect all of them to be executed. I would also expect the same for the after.

**Comment by jgebal** [comment]
About the savepoints.
Should we set a savepoint on each level of the `before` and rollback to it on each `after`?
It sounds reasonable to allow savepoints everywhere.

**Comment by jgebal** [comment]
@pazus I think we should extract the not yet implemented annotations into a separate task and close this one.
Seems like most of it is one and we are on track with the approach we want to follow.

**Comment by Pazus** [comment]
Agree

**Comment by Pazus** [comment]
Moved discussion to #182
